### ISSUE TITLE
security enhanced mode requires initialization and then restart

### ISSUE THREAD

**Issue by gregwebs** [pingcap/tidb#31201]
## Enhancement

Make it possible to use SEM without restarting TiDB when bringing up a new TiDB cluster.

### Problem

RESTRICTED_* permissions cannot be set by a user with `SUPER` or `ALL PRIVILEGES`.
Once `security.enable-sem = true` is set on a TiDB instance, it is not possible to grant `RESTRICTED_*`, even on a separate tidb instance with `security.enable-sem = false` and/or `skip-grant-table = true`

This requires the following workflow
1) start tidb with `enable-sem = false`
2) run SQL that grants `RESTRICTED_VARIABLES_ADMIN`, etc
3) restart tidb with `enable-sem = true`

This also means that it is important to grant all RESTRICTED_ permissions to a user before enabling sem.
Otherwise a user must temporarily reduce their security by using the above workflow.

This also means that anytime tidb introduces a new RESTRICTED_ permission, it requires users to temporarily reduce their security to grant that permission.

## Solution

There are different ways to handle this bootstrapping. There is only one I can think of that doesn't have very obvious drawbacks.

#### Configure a user with access to RESTRICTED privileges

```
security.sem-user = "root"
```

This user is given the ability to grant RESTRICTED privileges or perhaps to use them. Once the RESTRICTED_ privileges are granted the configuration may be disabled (improves security if they don't need all of the RESTRICTED privileges).

@morgo

**Comment by SunRunAway** [comment]
> This also means that anytime tidb introduces a new RESTRICTED_ permission, it requires users to temporarily reduce their security to grant that permission.

This can be done by joining a temporary tidb-server with SEM=false.

**Comment by SunRunAway** [comment]
Making sem-user be able to grant privileges may introdcues extra attacking point.
before: the attacker cannot read customer's data even if they can access sem user
after: the attacker can read customer's data if they can access sem user by grant a new account.

**Comment by gregwebs** [comment]
> This can be done by joining a temporary tidb-server with SEM=false.

Have you tested this? In my testing there is no way to get this to work. If it worked it would be a huge security hole.

**Comment by gregwebs** [comment]
One of the problems with trying to join a separate TiDB was that the dynamic privileges do not get enabled by the parser so there is a parse error on the RESTRICTED privileges.

**Comment by morgo** [comment]
> One of the problems with trying to join a separate TiDB was that the dynamic privileges do not get enabled by the parser so there is a parse error on the RESTRICTED privileges.

This shouldn't be the case. You might be using an earlier version of TiDB? Because SEM is instance scoped, no TiDB servers share with the cluster what their configuration is.

The current behavior is an intentional restriction. In MySQL, there are various settings which are intentionally non-dynamic because they could create an attack surface area.

**Comment by SunRunAway** [comment]
So restarting TiDB seems to be the best and security way.
both joining a non-SEM tidb instance and making an SEM user who can grant privileges expand the attacking point.

**Comment by gregwebs** [comment]
>  You might be using an earlier version of TiDB? Because SEM is instance scoped, no TiDB servers share with the cluster what their configuration is.

No, I am using the 5.3 release. The [source code comments indicate that this is a global variable](https://github.com/pingcap/tidb/blob/master/sessionctx/variable/tidb_vars.go#L618). @morgo am I reading the source code correctly? Regardless of what the docs say (they agree with you), I can assure you that joining a non SEM instance doesn't work (I have tested this extensively), and I only know how to explain this as there being some kind of shared state going on.

**Comment by gregwebs** [comment]
> both joining a non-SEM tidb instance and making an SEM user who can grant privileges expand the attacking point.

This is incorrect. Currently the attacking point is that anyone with SUPER can attack when SEM is disabled. And it must be disabled temporarily to satisfy certain workflows including the initial bootstrap.

With the SEM user approach only a single named SEM user can attack as opposed to any user with SUPER. In this design, `enable-sem` will be set by users to always be true, including during bootstrap, and the `sem-user` config would be designed to be temporarily changed just as `enable-sem` must be temporarily changed.

**Comment by morgo** [comment]
> No, I am using the 5.3 release. The [source code comments indicate that this is a global variable](https://github.com/pingcap/tidb/blob/master/sessionctx/variable/tidb_vars.go#L618). @morgo am I reading the source code correctly? Regardless of what the docs say (they agree with you), I can assure you that joining a non SEM instance doesn't work (I have tested this extensively), and I only know how to explain this as there being some kind of shared state going on.

OK, I'll take a look tomorrow and try and reproduce.

> This is incorrect. Currently the attacking point is that anyone with SUPER can attack when SEM is disabled. And it must be disabled temporarily to satisfy certain workflows including the initial bootstrap.

You can bring this up with the security team if you like. I don't have a horse in this race, but I'm not going to change it unless they tell me it's okay to do.

**Comment by gregwebs** [comment]
> You can bring this up with the security team if you like. I don't have a horse in this race, but I'm not going to change it unless they tell me it's okay to do.

Are we going to add new RESTRICTED_ privileges? I think the current design is untenable in that case because it requires SEM to temporarily be disabled to use such a new privilege. If we don't introduce any new RESTRICTED_ privileges then the current design is okay.

**Comment by morgo** [comment]
> OK, I'll take a look tomorrow and try and reproduce.

I tested this today, it works fine:

1. Setup the first server with SEM enabled in configuration.
2. Setup a second tidb-server with: `/home/morgo/go/src/github.com/morgo/tidb/bin/tidb-server --store=tikv --path="127.0.0.1:2379" -P 4001 -status 10081` (my path to a tidb-server binary).
3. Connect to the new TiDB server with: `mysql -P 4001 -h 127.0.0.1`

In the original session, I can create a user but I can't grant a restricted privilege:

```
tidb> show grants;
+-------------------------------------------------------------+
| Grants for User                                             |
+-------------------------------------------------------------+
| GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' WITH GRANT OPTION |
+-------------------------------------------------------------+
1 row in set (0.00 sec)

tidb> CREATE user ted;
Query OK, 0 rows affected (0.05 sec)

tidb> grant restricted_tables_admin on *.* to ted;
ERROR 1227 (42000): Access denied; you need (at least one of) the GRANT OPTION privilege(s) for this operation
tidb> show global variables like '%security';
+-------------------------------+-------+
| Variable_name                 | Value |
+-------------------------------+-------+
| tidb_enable_enhanced_security | ON    |
+-------------------------------+-------+
1 row in set (0.00 sec)
```

But in the new server I can:

```
tidb> show global variables like '%security';
+-------------------------------+-------+
| Variable_name                 | Value |
+-------------------------------+-------+
| tidb_enable_enhanced_security | OFF   |
+-------------------------------+-------+
1 row in set (0.01 sec)

tidb> grant restricted_tables_admin on *.* to ted;
Query OK, 0 rows affected (0.01 sec)

```

Back on the original server:

```
tidb> SHOW GRANTS FOR ted;
+---------------------------------------------------+
| Grants for ted@%                                  |
+---------------------------------------------------+
| GRANT USAGE ON *.* TO 'ted'@'%'                   |
| GRANT RESTRICTED_TABLES_ADMIN ON *.* TO 'ted'@'%' |
+---------------------------------------------------+
2 rows in set (0.00 sec)
```

**Comment by gregwebs** [comment]
@morgo thanks for testing. Let me compare this to my k8s-based setup and see what the difference is.

BTW, is this the designed behavior that I can rely on? It seems like a big security hole.
